### DISCUSSION TITLE
Marking boundary attributes of structured grid

### DISCUSSION THREAD

**Discussion by s769** [pyvista/pyvista#6800] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

If I have a structured grid of a unit cube, how can I "tag" each boundary face with a separate value so that I can use it for finite element analysis?

```python
import pyvista as pv
import numpy as np

Nx, Ny, Nz = 20,10, 5
x = np.linspace(0,1, Nx+1)
z = np.linspace(0,1, Ny+1)
y = np.linspace(0,1, Nz+1)
x, y, z = np.meshgrid(x, y, z)

grid = pv.StructuredGrid(x, y, z)
```

### Description

I know you can add point data based on the point coordinates, but this becomes ambiguous for points that lie on the intersection of 2 faces of the unit cube. If I use cell data, it will apply to the entire 3D cell instead of the 2D face. This also has the problem of ambiguity when a cell is on the intersection of 2 faces of the unit cube.

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Thu Nov 14 16:10:00 2024 Central Standard Time

                  OS : Windows
              CPU(s) : 12
             Machine : AMD64
        Architecture : 64bit
                 RAM : 31.7 GiB
         Environment : IPython
          GPU Vendor : Intel
        GPU Renderer : Intel(R) UHD Graphics 630
         GPU Version : 4.5.0 - Build 31.0.101.2121
    MathText Support : False

  Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64
  bit (AMD64)]

             pyvista : 0.44.1
                 vtk : 9.2.6
               numpy : 2.1.3
          matplotlib : 3.9.2
              scooby : 0.7.1
               pooch : v1.6.0
              pillow : 9.5.0
             imageio : 2.35.1
             IPython : 8.10.0
          ipywidgets : 8.0.4
               scipy : 1.14.1
                tqdm : 4.66.5
              meshio : 5.3.5
          jupyterlab : 3.6.1
jupyter_server_proxy : 3.2.2
        nest_asyncio : 1.5.6
--------------------------------------------------------------------------------
```

**Comment by darikg** [comment]
I'm pretty sure you'd have to maintain some sort of "dual mesh" in parallel, e.g. a pointset with 6 * nx * ny * nz points.

**Comment by user27182** [comment]
What about using multi-component scalars? E.g. use cell data, but instead of a single value stored for each cell, store 6 values, one for each face. You can adopt a convention like `(-x, +x, -y, +y, -z, +z)` for the data, such that component zero is for the `-x` face, etc. (This is the same convention used by `mesh.bounds`).

The visualization isn't great, since you still have cubic cells showing even though the values are tied to faces, but this way you can filter values per-face, e.g. using `extract_values`

``` python
import pyvista as pv
import numpy as np

Nx, Ny, Nz = 2,2, 2
x = np.linspace(0,1, Nx+1)
z = np.linspace(0,1, Ny+1)
y = np.linspace(0,1, Nz+1)
x, y, z = np.meshgrid(x, y, z)

grid = pv.StructuredGrid(x, y, z)

# Add a length-6 vector of cell data for each cell
grid.cell_data['data'] = np.reshape(range(grid.n_cells*6), (grid.n_cells,6))

# Extract cells which have a value between 0-25 for the negative x face
component = 0  # negative x component
x_neg = grid.extract_values(ranges=[0,25], component_mode=component)
x_neg.plot(opacity=0.5, show_edges=True, component=component)
```
